File: golang/spamhaus.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"strings"
)

func reverse(ss []string) {
	last := len(ss) - 1
	for i := 0; i < len(ss)/2; i++ {
		ss[i], ss[last-i] = ss[last-i], ss[i]
	}
}

type ipRec struct {
	Ip string   `json:"ip"`
	Code string `json:"code"`
	Zone string `json:"zone"`
	Desc string `json:"desc"`
}

func jsonRec(rec ipRec) string {
	js, err := json.Marshal(rec)
	if err != nil {
		return("==>{}")
	}
	return(string(js))
}

func humanRec(rec ipRec) string {
	return(rec.Ip + " " + rec.Zone + " " + rec.Desc)
}

func isValidIPv4(ip string) bool {
	parsed := net.ParseIP(ip)
	if (parsed == nil) { return(false) }
	return(parsed.To4() != nil)
}

func classifySpamhaus(arg string) []ipRec {
	
	out := []ipRec{}
	
	if (!isValidIPv4(arg)) {
		out = append(out, ipRec{arg, "NA", "NA", "Not a valid IPv4 address"})
		return(out)
	}
	
	octets := strings.Split(arg, ".")
	reverse(octets)
	
	ip := (strings.Join(octets, ".")) + ".zen.spamhaus.org"
	
	ips, err := net.LookupIP(ip)
	
	if err != nil {
		out = append(out, ipRec{arg, "NA", "nbl", "Not on any Spamhaus blocklist"})
	} else {
		for _, res := range ips {
			switch res.String() {
				case  "127.0.0.2": out = append(out, ipRec{ arg, res.String(), "SBL", "Spamhaus SBL Data"})
				case  "127.0.0.3": out = append(out, ipRec{ arg, res.String(), "SBL", "Spamhaus SBL CSS Data"})
				case  "127.0.0.4": out = append(out, ipRec{ arg, res.String(), "XBL", "CBL Data"})
				case  "127.0.0.9": out = append(out, ipRec{ arg, res.String(), "SBL", "Spamhaus DROP/EDROP"})
				case "127.0.0.10": out = append(out, ipRec{ arg, res.String(), "PBL", "ISP Maintained"})
				case "127.0.0.11": out = append(out, ipRec{ arg, res.String(), "PBL", "Spamhaus Maintained"})
				default: out = append(out, ipRec{ arg, "nbl", "NA", "Not on any Spamhaus blocklist"})
			}
		}
	}
	
	return(out)
	
}
	
func main() {
	
	args := os.Args[1:]
	
	// input on stdin
	if (len(args) == 0) || ((len(args) == 1) && (args[0] == "-")) {
		
		stdin := bufio.NewScanner(os.Stdin)
		for stdin.Scan() {
			txt := stdin.Text()
			for _, out := range classifySpamhaus(txt) {
				fmt.Println(jsonRec(out))
			}
		}
		
	} else {
			
		for _, arg := range args {
			for _, out := range classifySpamhaus(arg) {
				fmt.Println(humanRec(out))
			}
		}
	}
		
}

File: R/spamhaus.R
#!env Rscript

library(httr)
library(ipaddress)
library(stringi)
library(gdns)

# get cmdline args to see if this is where ips are
args <- commandArgs(trailingOnly = TRUE)

is_stdin <- FALSE

# if no cmdline args assume stdin
if ((length(args) == 0) | (args[1] == "-")) { # read from stdin
  ips <- readLines(file("stdin"), warn = FALSE)
  is_stdin <- TRUE
} else {
  ips <- args
}

ips <- unique(ips)

if (!length(ips)) exit(0) # bail if nothing to do

# validate the IPs
validate <- suppressWarnings(ip_address(ips)) 
bad <- ips[is.na(validate) | (!is_ipv4(validate))]

if (length(bad)) {
  warning(scales::comma(length(bad), 1), " invalid IPv4 addresses in input.")
}

# the good stuff (if any)
ok  <- ip_address(ips[is_ipv4(validate)])
if (length(ok) == 0) exit(0) # bail if no ips

# turn them into what spamhaus needs
spamhaus_ptr <- sub("in-addr.arpa", "zen.spamhaus.org", reverse_pointer(ok), fixed = TRUE)

# do the thing
lapply(
  spamhaus_ptr,
  curl::nslookup, ipv4_only = TRUE, multiple = TRUE, error = FALSE
) -> res

# if any answers are empty, mark them as not on the spamhaus block list
res[lengths(res) == 0] <- "nbl"
res <- unlist(setNames(res, ok))

# make a tidy data frame
merge(
  data.frame(
    ip = names(res),
    code = as.character(res)
  ),
  data.frame( # https://www.spamhaus.org/faq/section/DNSBL%20Usage#200
    code = c("nbl", "127.0.0.2", "127.0.0.3", "127.0.0.4", "127.0.0.9", "127.0.0.10", "127.0.0.11"),
    zone = c(NA_character_, "SBL", "SBL", "XBL", "SBL", "PBL", "PBL"),
    desc = c("Not on any Spamhaus blocklist", "Spamhaus SBL Data", "Spamhaus SBL CSS Data", "CBL Data", "Spamhaus DROP/EDROP Data", "ISP Maintained", "Spamhaus Maintained")
  ),
  by = "code",
  all.x = TRUE
) -> res

# if we got stuff via stdin, chances are good we want programmatic output
# so return ndjson, otherwise print the data frame
if (is_stdin) {
  jsonlite::stream_out(res[,c("ip", "code", "zone", "desc")], stdout(), verbose = FALSE)
} else {
  print(res[,c("ip", "code", "zone", "desc")]), max = nrow(res))
}
